fix(File): Handle file_path is None

tundebabzy opened the pull request:
In `validate_file_size`, if `file_path` is invalid i.e `None`,
set `file_size` to 0

<!--

Some key notes before you open a PR:

 1. Select which branch should this PR be merged in?
 2. PR name follows [convention](http://karma-runner.github.io/4.0/dev/git-commit-msg.html)
 3. All tests pass locally, UI and Unit tests
 4. All business logic and validations must be on the server-side
 5. Update necessary Documentation
 6. Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes


Also, if you're new here

- Documentation Guidelines => https://github.com/frappe/erpnext/wiki/Updating-Documentation

- Contribution Guide => https://github.com/frappe/frappe/blob/develop/.github/CONTRIBUTING.md

- Pull Request Checklist => https://github.com/frappe/erpnext/wiki/Pull-Request-Checklist

-->

> Please provide enough information so that others can review your pull request:

closes #15550 . Please backport to older versions if PR is accepted as it affects older versions. The bug was detected in version-12

<!-- You can skip this if you're fixing a typo or updating existing documentation -->

> Explain the **details** for making this change. What existing problem does the pull request solve?

This ensures `validate_file_size` does not throw an Exception when there is no backup file to validate i.e when `get_latest_backup_file` returns `None` as the `file_path`. 

<!-- Example: When "Adding a function to do X", explain why it is necessary to have a way to do X. -->

> Screenshots/GIFs

<!-- Add images/recordings to better visualize the change: expected/current behviour -->

tundebabzy committed:
fix: S3 automatic backup fails when backup folder is empty

hasnain2808 commented on frappe/integrations/offsite_backup_utils.py:
```diff
@@ -66,8 +66,8 @@ def get_latest_backup_file(with_files=False):
 
 
 def get_file_size(file_path, unit):
-	if not unit:
```
Let's not remove this validation

hasnain2808 commented on frappe/integrations/offsite_backup_utils.py:
```diff
@@ -66,8 +66,8 @@ def get_latest_backup_file(with_files=False):
 
 
 def get_file_size(file_path, unit):
-	if not unit:
```
This validation prevents sending '' for unit
Let's not remove it

tundebabzy replied:
Hi @hasnain2808

I actually forget to remove that conditional since I had concluded that a better path will be to not bother finding the file size of a file whose path is `None` or empty string.

hasnain2808 replied:
@tundebabzy 
how about https://github.com/frappe/frappe/pull/15552/files#r788581217

hasnain2808 replied:
We need a default value it will fail [here](https://github.com/frappe/frappe/pull/15552/files#diff-d6b9a7d94568ce928c5d605d5d1dd25c5b5cddbf6bc486a471bae00fb039d1b7L76) otherwise
I think this is a better way
https://github.com/frappe/frappe/pull/15552/files#r788581217

tundebabzy committed:
Merge branch 'develop' into 15550

tundebabzy committed:
Removed conditional that returns early when
`file_path` argument is invalid i.e `None` or empty string.

This means though that the function will throw an exception if it
receives such invalid input so devs should check their inputs before
calling the function

hasnain2808 commented:
@tundebabzy @shadrak98 

The fix for both of these problems(the current PR and https://github.com/frappe/frappe/pull/15566) is probably the same which is verifying if the link actually points to a valid file. File doctype/ file uploader should do this validation before accepting a file the backup services can reuse the functions to make sure direct db updates were correct.

tundebabzy commented:
Hi @hasnain2808, I'm not sure I understand your comment. What link are you referring to?

The function (`validate_file_size`) does not have anything to do with the File doctype or uploader. `validate_file_size` literally tries to get the latest database file it can find and sets `frappe.flags.create_new_backup` to `False` if the file size is greater than 1. It literally wants to know what the value of `frappe.flags.create_new_backup` should be set to. Nothing else.

The code calling this function in the S3 Integration looks like
```(python)
def take_backups_s3(retry_count=0):
	try:
		validate_file_size()
		backup_to_s3()
		send_email(True, "Amazon S3", "S3 Backup Settings", "notify_email")
	except JobTimeoutException:
            .....
```

hasnain2808 commented on frappe/integrations/offsite_backup_utils.py:
```diff
@@ -66,9 +66,6 @@ def get_latest_backup_file(with_files=False):
 
 
 def get_file_size(file_path, unit):
```
```suggestion
def get_file_size(file_path, unit='MB'):
```
https://github.com/frappe/frappe/pull/15552/files#r788583827

hasnain2808 commented:
> Hi @hasnain2808, I'm not sure I understand your comment. What link are you referring to?
> 
> The function (`validate_file_size`) does not have anything to do with the File doctype or uploader. `validate_file_size` literally tries to get the latest database file it can find and sets `frappe.flags.create_new_backup` to `False` if the file size is greater than 1. It literally wants to know what the value of `frappe.flags.create_new_backup` should be set to. Nothing else.
> 
> The code calling this function in the S3 Integration looks like
> 
> ```
> def take_backups_s3(retry_count=0):
> 	try:
> 		validate_file_size()
> 		backup_to_s3()
> 		send_email(True, "Amazon S3", "S3 Backup Settings", "notify_email")
> 	except JobTimeoutException:
>             .....
> ```

Thanks, @tundebabzy for pointing it out. I mixed this fix with a different issue we are working on.
I have added a small change other than that I think we are good to go

tundebabzy committed:
Update frappe/integrations/offsite_backup_utils.py

Co-authored-by: Mohammad Hasnain Mohsin Rajan <hasnain2808@gmail.com>

hasnain2808 commented:
@Mergifyio backport version-13-hotfix

tundebabzy commented:
@hasnain2808 Thanks for approving the PR. Is this going into v12 as well? The bug also exists there

tundebabzy commented:
@Mergifyio backport version-12-hotfix

tundebabzy commented:
@mergify backport version-12-hotfix
